### PR TITLE
perf: faster ffmpeg decoder

### PR DESCRIPTION
**Pull request by justusmattern27**

### CODE REVIEW

File: packages/vite-plugin/src/partials/ffmpegBridge.ts [hunk 1]
```diff
@@ -85,6 +86,89 @@ export function ffmpegBridgePlugin({output}: ExporterPluginConfig): Plugin {
           }
         },
       );
+
+      server.middlewares.use(
```
Comment (hkonsti) [inline packages/vite-plugin/src/partials/ffmpegBridge.ts:90]: There is a lot of repeated code here 😅. Claude gave me this which looks like a good starting point:
```ts
import {
  FFmpegExporterServer,
  FFmpegExporterSettings,
  VideoFrameExtractor,
  generateAudio,
  mergeMedia,
} from '@revideo/ffmpeg';
import {existsSync, unlinkSync} from 'fs';
import type {Plugin, WebSocketServer} from 'vite';

// ... existing code ...

export function ffmpegBridgePlugin({output}: ExporterPluginConfig): Plugin {
  return {
    name: 'revideo/ffmpeg',

    configureServer(server) {
      const ffmpegBridge = new FFmpegBridge(server.ws, {output});

      const handlePostRequest = async (req, res, handler) => {
        if (req.method !== 'POST') {
          res.statusCode = 405;
          res.end('Method Not Allowed');
          return;
        }

        const body = await new Promise((resolve, reject) => {
          let data = '';
          req.on('data', chunk => (data += chunk));
          req.on('end', () => resolve(data));
          req.on('error', reject);
        });

        try {
          const result = await handler(JSON.parse(body));
          res.statusCode = 200;
          if (result) {
            res.setHeader('Content-Type', 'application/json');
            res.end(JSON.stringify(result));
          } else {
            res.end('OK');
          }
        } catch (e) {
          console.error(e);
          res.statusCode = 500;
          res.end('Internal Server Error');
        }
      };

      server.middlewares.use('/audio-processing/generate-audio', (req, res) =>
        handlePostRequest(req, res, async ({tempDir, assets, startFrame, endFrame, fps}) =>
          generateAudio({outputDir: output, tempDir, assets, startFrame, endFrame, fps})
        )
      );

      server.middlewares.use('/audio-processing/merge-media', (req, res) =>
        handlePostRequest(req, res, async ({outputFilename, tempDir}) =>
          mergeMedia(outputFilename, output, tempDir)
        )
      );

      server.middlewares.use('/revideo-ffmpeg-decoder/video-frame', (req, res) =>
        handlePostRequest(req, res, async (data) => {
          const {frame, width, height} = await ffmpegBridge.handleDecodeVideoFrame(data);
          res.setHeader('X-Frame-Width', width.toString());
          res.setHeader('X-Frame-Height', height.toString());
          res.setHeader('Content-Type', 'application/octet-stream');
          res.end(Buffer.from(frame as Buffer));
        })
      );

      server.middlewares.use('/revideo-ffmpeg-decoder/finished', (req, res) =>
        handlePostRequest(req, res, () => ffmpegBridge.handleRenderFinished())
      );

      server.middlewares.use('/revideo-ffmpeg-decoder/download-video-chunks', (req, res) =>
        handlePostRequest(req, res, async (videoDurations) => {
          const downloadedPaths = await ffmpegBridge.handleDownloadVideoChunks(videoDurations);
          return {success: true, paths: downloadedPaths};
        })
      );
    },
  };
}
```

File: packages/core/src/exporter/download-videos.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import {AssetInfo} from '../app';
+
+export async function download(assets: AssetInfo[][]): Promise<void> {
+  const videoRanges: Map<string, {start: number; end: number}> = new Map();
+
+  assets.forEach(frameAssets => {
```
Comment (hkonsti) [inline packages/core/src/exporter/download-videos.ts:6]: When ended up with nested if/elses like this please consider rewriting it to make it more readable.
```ts
  assets.forEach(frameAssets => {
    frameAssets.forEach(asset => {
      if (asset.type !== 'video' || asset.decoder !== 'ffmpeg') {
        return;
      }

      if (videoRanges.has(asset.src)) {
          const range = videoRanges.get(asset.src)!;
          range.start = Math.min(range.start, asset.currentTime);
          range.end = Math.max(range.end, asset.currentTime);
          return;
      }

      videoRanges.set(asset.src, {
        start: asset.currentTime,
        end: asset.currentTime,
      });
    });
  });
  ```

File: packages/core/src/exporter/download-videos.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    throw new Error(`HTTP error! status: ${response.status}`);
+  }
+
+  const result = await response.json();
+
+  if (!result.success) {
+    throw new Error(`Error downloading video chunks: ${result.error}`);
+  }
+
+  console.log('finished downloading');
```
Comment (hkonsti) [inline packages/core/src/exporter/download-videos.ts:57]: We should probably come up with a policy on console.logs in the future. I think it's fine to leave this in for now but it makes Revideo seem more immature than it is when it emits random unstyled logs. We could maybe look at how other libraries handle this.

Same goes for error handling too. Right now we just randomly throw here and there but never really handle anything.

File: packages/2d/src/lib/utils/video/ffmpeg-client.ts [hunk 4]
```diff
@@ -28,37 +12,40 @@ export class ImageCommunication {
...
-          id: id,
-          filePath: src,
-          startTime: time,
-          duration,
-          fps,
-        },
-      });
-    });
+    const imageBitmap = await createImageBitmap(imageData);
+    return imageBitmap;
```
[the remark this answers is not in the document]
Comment (justusmattern27) [reply packages/2d/src/lib/utils/video/ffmpeg-client.ts:49]: done

File: packages/ffmpeg/src/video-frame-extractor.ts [hunk 5]
```diff
@@ -49,36 +55,91 @@ export class VideoFrameExtractor {
...
+    )?.localPath as string;
+    this.startTimeOffset = VideoFrameExtractor.downloadedVideoMap.get(filePath)
+      ?.startTimeOffset as number;
 
     this.startTime = startTime;
     this.duration = duration;
     this.toTime = this.getEndTime(this.startTime);
     this.fps = fps;
 
     if (this.startTime >= this.duration) {
```
[the remark this answers is not in the document]
Comment (justusmattern27) [reply packages/ffmpeg/src/video-frame-extractor.ts:69]: done

File: packages/core/src/exporter/download-videos.ts (hunk 3 above)
Comment (justusmattern27) [reply packages/core/src/exporter/download-videos.ts:57]: fully agree

**Review by hkonsti — APPROVED** [review]
nice!!!

**Pull request merged**